packages/ag-charts-community/src/components/toolbar/toolbar.ts

```diff
@@ -0,0 +1,177 @@
+import { BaseProperties } from '../../util/properties';
+import type { ButtonWidget as BaseButtonWidget } from '../../widget/buttonWidget';
+import { NativeWidget } from '../../widget/nativeWidget';
+import { ToolbarWidget } from '../../widget/toolbarWidget';
+import { ToolbarButtonWidget, type ToolbarButtonWidgetOptions } from './toolbarButtonWidget';
+
+const BUTTON_ACTIVE_CLASS = 'ag-charts-toolbar__button--active';
+
+export interface ToolbarButtonOptions extends ToolbarButtonWidgetOptions {
+    section?: string;
+}
+
+export abstract class BaseToolbar<
+    ButtonOptions extends ToolbarButtonOptions,
+    ButtonWidget extends ToolbarButtonWidget,
+> extends ToolbarWidget {
+    public horizontalSpacing = 10;
+    public verticalSpacing = 10;
+
+    private readonly buttonWidgets: Array<ButtonWidget> = [];
+
+    constructor(
+        protected readonly ctx: ModuleContext,
+        private readonly onButtonPress: (button: ButtonOptions & { index: number; sourceEvent: MouseEvent }) => void,
+        private readonly onDragStart?: (sourceEvent: MouseEvent, element: HTMLElement) => void
+    ) {
+        super();
+
+        const element = this.getElement();
+        element.classList.add('ag-charts-toolbar');
+
+        this.createDragHandle();
+    }
+
+    public updateButtons(buttons: Array<ButtonOptions>) {
+        const { buttonWidgets } = this;
+
+        for (const [index, button] of buttons.entries()) {
+            const buttonWidget = this.buttonWidgets.at(index) ?? this.createButton(index, button);
+            buttonWidget.update(button);
+        }
+
+        for (let index = buttons.length; index < buttonWidgets.length; index++) {
+            const button = this.buttonWidgets.at(index);
+            // this.toolbar.removeChild(button); // TODO
+            button?.destroy();
+        }
+
+        this.refreshButtonClasses();
+    }
+
+    public updateButtonByIndex(index: number, button: ButtonOptions) {
+        this.buttonWidgets.at(index)?.update(button);
+    }
+
+    public clearActiveButton() {
+        for (const button of this.buttonWidgets) {
+            button.getElement().classList.toggle(BUTTON_ACTIVE_CLASS, false);
+        }
+    }
+
+    public toggleActiveButtonByIndex(index: number) {
+        for (const [buttonIndex, button] of this.buttonWidgets.entries()) {
+            button.getElement().classList.toggle(BUTTON_ACTIVE_CLASS, index != null && index === buttonIndex);
+        }
+    }
+
+    public toggleButtonEnabledByIndex(index: number, enabled: boolean) {
+        const buttonWidget = this.buttonWidgets.at(index);
+        if (!buttonWidget) return;
+        buttonWidget.getElement().ariaDisabled = (!enabled).toString();
+    }
+
+    public toggleSwitchCheckedByIndex(index: number, checked: boolean) {
+        const buttonWidget = this.buttonWidgets.at(index);
+        if (!buttonWidget) return;
+        buttonWidget.getElement().ariaChecked = checked.toString();
```

Comment:
It seemed like the `ButtonWidget` probably shouldn't have this since there's the `SwitchWidget` which extends it and adds `setChecked()`?

So I added it to `ToolbarButtonWidget` instead. Did think about making that extend `SwitchWidget` but then it has to remove the role in the constructor, which seemed a bit wrong.

Ideally it would be a mixture of buttons and switches but that's more complex than I want to go for now.

Reply:
🤔 Hmm. Yes, you are right. I was thinking of toggle-buttons, but that uses the `aria-pressed` attribute (https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role#associated_aria_roles_states_and_properties).

We want `SwitchWidget.setChecked` instead of `ButtonWidget.setChecked` to avoid misuse.